src/sage/schemes/weighted_projective/weighted_projective_point.py

```diff
@@ -0,0 +1,297 @@
+        return "({})".format(" : ".join(map(repr, self._coords)))
+
+    def _richcmp_(self, other, op):
+        """
+        Test the weighted projective equality of two points.
+
+        INPUT:
+
+        - ``other`` -- a point on weighted projective space
+
+        OUTPUT:
+
+        Boolean
+
+        EXAMPLES::
+
+            sage: WP = WeightedProjectiveSpace([3, 4, 5], QQ)
+            sage: u = abs(QQ.random_element()) + 1 / 4 # nonzero
+            sage: P, Q = WP(2, 3, 4), WP(2 * u^3, 3 * u^4, 4 * u^5)
+            sage: P == Q
+            True
+            sage: P != Q
+            False
+
+        ::
+
+            sage: P, Q = WP(2, 3, 4), WP(2, 3, 5)
+            sage: P < Q
+            True
+            sage: P <= Q
+            True
+            sage: P > Q
+            False
+            sage: P >= Q
+            False
+        """
+        assert isinstance(other, SchemeMorphism_point)
+
+        space = self.codomain()
+        if space is not other.codomain():
+            return op == op_NE
+
+        if op in [op_EQ, op_NE]:
+            weights = space._weights
+            # (other[i] / self[i])^(1 / weight[i]) all equal
+            prod_weights = prod(weights)
+            # check weights
+            bw = weights == other.codomain()._weights
+            if bw != (op == op_EQ):
+                return False
+
+            # check zeros
+            b1 = all(c1 == c2
+                     for c1, c2 in zip(self._coords, other._coords)
+                     if c1 == 0 or c2 == 0)
+            if b1 != (op == op_EQ):
+                return False
+
+            # check nonzeros
+            ratio = [(c1 / c2) ** (prod_weights // w)
+                     for c1, c2, w in zip(self._coords, other._coords, weights)
+                     if c1 != 0 and c2 != 0]
+            r0 = ratio[0]
+            b2 = all(r == r0 for r in ratio)
+            return b2 == (op == op_EQ)
+
+        return richcmp(self._coords, other._coords, op)
+
+    def __hash__(self):
+        """
+        Compute the hash value of this point.
+
+        We attempt to normalise the coordinates of this point over the field of
+        fractions of the base ring. If this is not possible, return the hash of
+        the parent. See :meth:`normalize_coordinates` for more details.
+
+        OUTPUT: Integer.
+
+        .. SEEALSO ::
+
+            :meth:`normalize_coordinates`
+
+        EXAMPLES::
+
+            sage: WP = WeightedProjectiveSpace([1, 3, 1], QQ)
+            sage: hash(WP(6, 24, 2)) == hash(WP(3, 3, 1))
+            True
+            sage: WP = WeightedProjectiveSpace([1, 3, 1], ZZ)
+            sage: hash(WP(6, 24, 2)) == hash(WP(3, 3, 1))
+            True
+        """
+        R = self.codomain().base_ring()
+        P = self.change_ring(FractionField(R))
+        P.normalize_coordinates()
+        if P._normalized:
+            return hash(tuple(P))
+        # if there is no good way to normalize return a constant value
```

Comment:
```suggestion
        # If there is no good way to normalize return the hash of the codomain.
        # This satisfies the requirement that elements which compare as equal have the same hash value.
```
Having `__hash__` return a constant is very suspicious, and not what this does.

Reply:
```
        # if there is no good way to normalize return a constant value
        return hash(self.codomain())
```

it is exactly what it does. It is constant in the sense that all points on that curve have equal hash value.

Reply:
When I read "constant value" I think of something like 5, not something that is constant for all points on the curve but differs across curves. Anyway, not a big deal.